### DISCUSSION TITLE
Hybrid rendering with ray tracing?

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#4180] [Ideas]
This is probably a little "out there", but has there been any thought of using some ray tracing?  This question is inspired by Apple's WWDC21 video [Explore hybrid rendering with Metal ray tracing](https://developer.apple.com/videos/play/wwdc2021/10150/).  I expect it couldn't be done with an OpenGL backend, but I think Vulkan has some ray tracing capability.

One possibility that Apple mentions is using ray tracing to decide whether a fragment is in light or shadow relative to a particular light.  With this approach, a shadow-casting point light would be no more difficult than a shadow-casting spot light, and it would be possible to implement area lights as well.

**Comment by romainguy** [comment]
You don't need Vulkan nor Metal to do this but it may not be as optimal. Our main constraint is not being able to rely on compute shaders. Note that raytracing makes hard shadows easy but soft shadows can be expensive. Also we can do area lights without it (but the shadows wouldn't be "correct").

**Reply by jwwalker** [reply]
Yes, I realize that there is a tradeoff between realism and computation.  The more physically accurate you want the shadows to be, the more rays you have to compute.

I'm curious to hear more about what you mean by "we can do area lights without it".

**Reply by romainguy** [reply]
The lighting part of area lights doesn't require ray tracing (at least for area lights that have analytical solutions).

**Reply by romainguy** [reply]
And a paper just dropped yesterday that shows how to compute [soft shadows from area lights analytically](https://drive.google.com/file/d/1q0q0zMp15syTtPW_T4DtMKxbGhz_SzG5/view?usp=sharing). It's based on LTC so prob. too expensive for mobile and most low/medium-end desktop configurations.
